src/anonymeter/evaluators/singling_out_evaluator.py

```diff
@@ -330,66 +402,131 @@ def multivariate_singling_out_queries(
 
     Returns
     -------
-    List[str]
+    List[pl.Expr]
         The singling out queries.
 
     """
-    so_queries = UniqueSinglingOutQueries()
-    medians = df.median(numeric_only=True)
+    unique_so_queries = UniqueSinglingOutQueries(max_size=n_queries)
+
+    medians = df.median()
+    medians_dict = medians.to_dicts()[0]
+    dtypes_dict = {col: df[col].dtype for col in df.columns}
 
     n_attempts = 0
+    if max_attempts is not None and batch_size > max_attempts:
+        batch_size = max_attempts
+
+    rng_to_use = rng_default if rng is None else rng
 
-    while len(so_queries) < n_queries:
+    while len(unique_so_queries) < n_queries:
         if max_attempts is not None and n_attempts >= max_attempts:
             logger.warning(
                 f"Reached maximum number of attempts {max_attempts} when generating singling out queries. "
-                f"Returning {len(so_queries.queries)} instead of the requested {n_queries}."
-                "To avoid this, increase the number of attempts or set it to ``None`` to disable "
-                "The limitation entirely."
+                f"Returning {len(unique_so_queries)} instead of the requested {n_queries}."
             )
-            return so_queries.queries
+            return unique_so_queries.queries
 
-        record = df.iloc[rng.integers(df.shape[0])]
-        columns = rng.choice(df.columns, size=n_cols, replace=False).tolist()
+        # Generate a batch of queries
 
-        query = _query_from_record(record=record, dtypes=df.dtypes, columns=columns, medians=medians)
+        # Pre-sample all random row indices
+        random_indices = rng_to_use.integers(
+            low=0, high=df.shape[0], size=batch_size
+        )
+
+        # Extract all records in bulk
+        records = df[random_indices].to_dicts()
+
+        # Pre-sample all column choices
+        selected_columns = [
+            rng_to_use.choice(df.columns, size=n_cols, replace=False).tolist()
+            for _ in range(batch_size)
+        ]
+
+        queries_batch = [
+            _query_from_record(
+                record,
+                dtypes_dict,
+                columns,
+                medians_dict,
+                rng,
+            )
+            for record, columns in zip(records, selected_columns)
+        ]
```

Comment:
NIT:

```suggestion
        queries_batch = [
            _query_from_record(
                records_batch,
                dtypes_dict,
                columns,
                medians_dict,
                rng,
            )
            for records_batch, columns in zip(records, selected_columns)
        ]
```

and consider naming the arguments.